docs/development/testing.md

```diff
@@ -1,26 +1,194 @@
+  - use `ExpectWithOffset` for making assertions in helper funcs like `expectSomethingWasCreated`: [example test](https://github.com/gardener/gardener/blob/2eb54485231408cbdbabaa49812572a07124364f/extensions/pkg/controller/controlplane/genericactuator/actuator_test.go#L732-L736)
+  - make sure to add additional descriptions to Gomega matchers if necessary (e.g. in a loop): [example test](https://github.com/gardener/gardener/blob/2eb54485231408cbdbabaa49812572a07124364f/test/e2e/shoot/internal/rotation/certificate_authorities.go#L89-L93)
+- introduce helper functions for assertions to make test more readable where applicable: [example test](https://github.com/gardener/gardener/blob/2eb54485231408cbdbabaa49812572a07124364f/test/integration/gardenlet/shootsecret/controller_test.go#L323-L331)
+- introduce custom matchers to make tests more readable where applicable: [example matcher](https://github.com/gardener/gardener/blob/2eb54485231408cbdbabaa49812572a07124364f/pkg/utils/test/matchers/matchers.go#L51-L57)
+- don't rely on accurate timing of `time.Sleep` and friends
+  - if doing so, CPU throttling in CI will make tests flaky, [example flake](https://github.com/gardener/gardener/issues/5410)
+  - use fake clocks instead, [example PR](https://github.com/gardener/gardener/pull/4569) 
+- use the same client schemes that are also used by production code to avoid subtle bugs/regressions: [example PR](https://github.com/gardener/gardener/pull/5469/commits/b0d9e2baeaee246c6eb466a181df35adca4a6ada)
+- use the debugger to ensure your test code is actually asserting the right thing
+- about overwriting global variables
+  - this is a common pattern (or hack?) in go for faking calls to external functions
+  - however, this can lead to races, when the global variable is used from a goroutine (e.g., the function is called)
+  - alternatively, set fields on structs (passed via parameter or set directly): this is not racy, as struct values are typically (and should be) only used for a single test case
+  - alternative to dealing with function variables and fields:
+    - add an interface, which your code depends on
+    - write a fake and a real implementation (similar to `clock.Clock.Sleep`)
+    - the real implementation calls the actual function (`clock.RealClock.Sleep` calls `time.Sleep`)
+    - the fake implementation does whatever you want it to do for your test (`clock.FakeClock.Sleep` waits until the test code advanced the time)
+- use constants in test code with care
+  - typically, you should not use constants from the same package as the tested code, instead use literals
+  - if the constant value is changed, tests using the constant will still pass, although the "specification" is not fulfilled anymore
+  - there are cases where it's fine to use constants, but keep this caveat in mind when doing so
+- creating sample data for tests can be a high effort
+  - if valuable, add a package for generating common sample data, e.g. Shoot/Cluster objects
```

Comment:
We could also mention about the "testdata" dir - it can be useful for some cases (helm charts, yaml manifests, etc). From https://pkg.go.dev/cmd/go/internal/test:
> The go tool will ignore a directory named "testdata", making it available
to hold ancillary data needed by the tests.

Example PR - https://github.com/gardener/gardener/pull/2140